content/blog/2020-11-11-self/index.md

```diff
@@ -0,0 +1,155 @@
+
+### Segregation
+
+This was a pretty interesting compiler optimization. Often, the heap is traversed and all references need to be processed. This is something the garbage collector does, something that needs to happen when an object is moved due to its size increasing, and something the "browser" does, which I believe is part of the Self IDE. Now, traversing through the heap looking for pointers becomes problematic when part of a byte-array happens to look like a memory address. Since byte-arrays are the only construct with this ambiguity in both Smalltalk and Self, Smalltalk implementations would traditionally have some extra overhead to check for byte-arrays as it processes the heap, so it knows to ignore pointer-like words which are part of a byte array.
+
+Instead, Chambers, Ungar, and Lee segregate the heap into sections where byte array objects can be allocated and sections where all other objects can be allocated. This way they avoid the overhead of being "on the watch" for byte arrays as they scan the heap.
+
+They also use two other tricks to speed up heap scans. First, at the end of the heap region, they place a dummy "sentinel" reference that gets the attention of the scanner. This way, the scanner only needs to check if it's within bounds when it sees a reference instead of every time it advances itself by one word. Similarly, a scanner often needs to capture the object which references another object. To minimize bookkeeping, every object begins with a special tag called a "mark". Then, when a reference is found, the scanner can just iterate backward to the closest mark to find the referencing object.
+
+Segregation, sentinels, and marks improve from the heap scanning speed of the fastest Smalltalk implementation by a factor of two.
+
+### Object Formats
+
+32-bit words in the slot contents can be one of three things: 30-bit integers, 30-bit floats, or 32-bit addresses. The two least significant bits encode which type of data the word represents. As far as the memory address is concerned, this type of encoding is an example of a [tagged pointer][taggedptr]. The last two bits are also used to differentiate the "mark" words described in the previous section, although these words are never placed inside slots.
+
+This section of the paper goes on to show how objects and maps are layed out in memory. I don't think there's anything too interesting or clever about this, but the kicker is that at the end the authors can demonstrate concretely that by using maps, their Self implementation is as space efficient as Smalltalk.
+
+## The Parser
```

Comment:
Seems like an odd title for a section that mostly focuses on the bytecode format & the JIT?